File: c_programs_and_containers/build_blocks/knowledge_base/sqlite3/data_structures/python/kb_link_mount_table.py
import sqlite3

class KB_Link_Mount_Table:
    def __init__(self, conn, cursor, database):
        """
        Initialize the LinkMountTable class.
        
        Args:
            conn: Database connection object
            cursor: Database cursor object
            database: Base name for the table (will be suffixed with '_link_mount')
        """
        self.conn = conn
        self.cursor = cursor
        self.base_table = database + "_link_mount"
        
        # Enable dictionary-like row access
        self.conn.row_factory = sqlite3.Row
        self.cursor = self.conn.cursor()
    
    def find_records_by_link_name(self, link_name, kb=None):
        """
        Find records by link_name, optionally filtered by parent_node_kb.
        
        Args:
            link_name (str): The link name to search for
            kb (str, optional): Knowledge base to filter by. If None, search all.
            
        Returns:
            list: List of matching records as dictionaries
        """
        if kb is None:
            query = f"""
                SELECT *
                FROM {self.base_table}
                WHERE link_name = ?
            """
            self.cursor.execute(query, (link_name,))
        else:
            query = f"""
                SELECT *
                FROM {self.base_table}
                WHERE link_name = ? AND knowledge_base = ?
            """
            self.cursor.execute(query, (link_name, kb))
        
        # Convert sqlite3.Row objects to dictionaries
        return [dict(row) for row in self.cursor.fetchall()]
    
    def find_records_by_mount_path(self, mount_path, kb=None):
        """
        Find records by mount_path, optionally filtered by knowledge_base.
        
        Args:
            mount_path (str): The mount path to search for
            kb (str, optional): Knowledge base to filter by. If None, search all.
            
        Returns:
            list: List of matching records as dictionaries
        """
        if kb is None:
            query = f"""
                SELECT *
                FROM {self.base_table}
                WHERE mount_path = ?
            """
            self.cursor.execute(query, (mount_path,))
        else:
            query = f"""
                SELECT *
                FROM {self.base_table}
                WHERE mount_path = ? AND knowledge_base = ?
            """
            self.cursor.execute(query, (mount_path, kb))
        
        # Convert sqlite3.Row objects to dictionaries
        return [dict(row) for row in self.cursor.fetchall()]
    
    def find_all_link_names(self):
        """
        Get all unique link names from the table.
        
        Returns:
            list: List of all unique link names
        """
        query = f"SELECT DISTINCT link_name FROM {self.base_table} ORDER BY link_name"
        self.cursor.execute(query)
        return [row["link_name"] for row in self.cursor.fetchall()]
    
    def find_all_mount_paths(self):
        """
        Get all unique mount paths from the table.
        
        Returns:
            list: List of all unique mount paths
        """
        query = f"SELECT DISTINCT mount_path FROM {self.base_table} ORDER BY mount_path"
        self.cursor.execute(query)
        return [row["mount_path"] for row in self.cursor.fetchall()]


# Example usage:
"""
import sqlite3

# Connect to database
conn = sqlite3.connect('your_database.db')
cursor = conn.cursor()

# Create instance
link_mount_table = KB_Link_Mount_Table(conn, cursor, "my_database")

# Use the methods
records = link_mount_table.find_records_by_link_name("example_link")
mount_records = link_mount_table.find_records_by_mount_path("root.docs.folder", kb="kb1")
all_links = link_mount_table.find_all_link_names()
all_mounts = link_mount_table.find_all_mount_paths()

# Don't forget to close the connection when done
conn.close()
"""

File: c_programs_and_containers/build_blocks/knowledge_base/sqlite3/data_structures/c/src/kb_link_mount_table.c
/*
 * kb_link_mount_table.c
 * Knowledge Base C Port — Link mount table queries
 */

#include "kb_link_mount_table.h"

#include <stdlib.h>

struct kb_link_mount_table {
    sqlite3 *db;
    char    *table_name;
};

kb_link_mount_table_t *kb_link_mount_table_create(sqlite3 *db,
                                                   const char *database)
{
    if (!db || !database) return NULL;

    kb_link_mount_table_t *lmt = (kb_link_mount_table_t *)calloc(1, sizeof(*lmt));
    if (!lmt) return NULL;

    lmt->db = db;
    lmt->table_name = kb_sprintf("%s_link_mount_table", database);
    if (!lmt->table_name) {
        free(lmt);
        return NULL;
    }
    return lmt;
}

void kb_link_mount_table_destroy(kb_link_mount_table_t *lmt)
{
    if (!lmt) return;
    free(lmt->table_name);
    free(lmt);
}

kb_error_t kb_link_mount_get_by_link_name(kb_link_mount_table_t *lmt,
                                           const char *link_name,
                                           kb_result_t *result)
{
    if (!lmt || !link_name || !result) return KB_ERR_NULL_ARG;
    kb_result_init(result);

    char *sql = kb_sprintf("SELECT * FROM %s WHERE link_name = ?",
                           lmt->table_name);
    if (!sql) return KB_ERR_NOMEM;

    kb_bind_param_t params[] = { KB_PARAM_TEXT(link_name) };
    kb_error_t err = kb_query_exec(lmt->db, sql, params, 1, result);
    free(sql);
    return err;
}

kb_error_t kb_link_mount_get_by_mount_path(kb_link_mount_table_t *lmt,
                                            const char *mount_path,
                                            kb_result_t *result)
{
    if (!lmt || !mount_path || !result) return KB_ERR_NULL_ARG;
    kb_result_init(result);

    char *sql = kb_sprintf("SELECT * FROM %s WHERE mount_path = ?",
                           lmt->table_name);
    if (!sql) return KB_ERR_NOMEM;

    kb_bind_param_t params[] = { KB_PARAM_TEXT(mount_path) };
    kb_error_t err = kb_query_exec(lmt->db, sql, params, 1, result);
    free(sql);
    return err;
}
